libnopegl/scripts/gen-doc.py

```diff
@@ -0,0 +1,192 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import json
+import sys
+from textwrap import dedent
+
+
+def _get_node_link(nodes, node_type):
+    params = nodes[node_type]
+    anchor = (params[1:] if isinstance(params, str) else node_type).lower()
+    return f"[{node_type}](#{anchor})"
+
+
+def _get_type_str(nodes, p) -> str:
+    node_type = p["type"]
+
+    choices = p.get("choices")
+    if choices:
+        type_str = f"[`{choices}`](#{choices}-choices)"
+    else:
+        type_str = f"[`{node_type}`](#parameter-types)"
+
+    node_types = p.get("node_types")
+    if node_types:
+        type_str += " (" + ", ".join(_get_node_link(nodes, x) for x in node_types) + ")"
+
+    return type_str
+
+
+def _get_default_str(p) -> str:
+    default = p.get("default")
+    if p["type"] == "select":
+        return f"`{default}`"
+    if p["type"] == "flags":
+        return f"`{default}`"
+    if p["type"] in ("f32", "f64"):
+        return f"`{default:g}`"
+    if p["type"] == "bool":
+        return "`unset`" if default < 0 else f"`{default}`"
+    if p["type"] in ("i32", "u32"):
+        return f"`{default}`"
+    if "ivec" in p["type"]:
+        return "(" + ",".join(f"`{x}`" for x in default) + ")"
+    if "vec" in p["type"]:
+        return "(" + ",".join(f"`{x:g}`" for x in default) + ")"
+    if "mat4" in p["type"]:
+        m = default
+        return "(" + " ".join(f"`{m[x]:g}`,`{m[x+1]:g}`,`{m[x+2]:g}`,`{m[x+3]:g}`" for x in range(0, 16, 4)) + ")"
+    if p["type"] == "str":
+        return f'"{default}"' if default is not None else ""
+    if p["type"] == "rational":
+        num, den = default
+        return f"`{num}/{den}`"
+    return ""
+
+
+def _get_node_params_block(nodes, name, params) -> str:
+    s = f"\n## {name}\n\n"
+
+    if not params:
+        return s
+
+    s += "Parameter | Flags | Type | Description | Default\n"
+    s += "--------- | ----- | ---- | ----------- | :-----:\n"
+    for param in params:
+        name = param["name"]
+        type = _get_type_str(nodes, param)
+        flags = "".join(f" [`{flag}`](#Parameter-flags)" for flag in param["flags"])
+        desc = param["desc"]
+        default = _get_default_str(param)
+        s += f"`{name}` | {flags} | {type} | {desc} | {default}\n"
+    s += "\n\n"
```

Comment:
Is "\n\n" intended here ? You use a single `\n` in  `_get_parameter_types_block` ?

Reply:
Yeah I wanted to preserve the current output, there are indeed 3 consecutive linebreaks happening in this situation

Reply:
Ok